.pre-commit-config.yaml

```diff
@@ -1,4 +1,11 @@
 repos:
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.0.287
+    hooks:
+      - id: ruff
+        args: [--fix, --exit-non-zero-on-fix]
```

Comment:
What does the autofix do for F811? I'm slightly suspicious of this, as you don't always fix this issue the same way. Sometimes you really do have two identical methods, and you just need to delete one. But sometimes one of the identical methods just needs to be renamed.

Reply:
Nothing for F811, because as you mention, it's not safe to delete one.

The rules with autofix are labelled with 🛠 at https://beta.ruff.rs/docs/rules/

I added it more for the future, but shall we remove it until then?

Reply:
I'm in favour of autofixes in general, but I think it's nice for people to be able to see easily from our config file which rules we expect to be autofixed where possible, and which rules we don't want autofixes for (even if they exist). We could maybe do something like this?

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 1d83c9414d..3cc332ab34 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -3,7 +3,7 @@ repos:
     rev: v0.0.287
     hooks:
       - id: ruff
-        args: [--fix, --exit-non-zero-on-fix]
+        args: [--exit-non-zero-on-fix]
         files: ^Lib/test/

   - repo: https://github.com/pre-commit/pre-commit-hooks
diff --git a/Lib/test/.ruff.toml b/Lib/test/.ruff.toml
index 1a381d84de..55619536d4 100644
--- a/Lib/test/.ruff.toml
+++ b/Lib/test/.ruff.toml
@@ -1,6 +1,10 @@
+fix = true
 select = [
     "F811",
 ]
+unfixable = ]
+    "F811",
+]
 extend-exclude = [
     # Failed to lint
     "badsyntax_pep3120.py",
```

Reply:
Hmm, I'm not sure about explicitly listing out all the fixer-less rules as unfixable, we'd be duplicating a lot. And some rules might grow a fixer later (e.g. it's not been implemented yet), and I think that would be the time to evaluate whether to disable it.

We're already being cautious and starting with just a single rule. And when we do have fixers, it is always up to the human to review and commit the changes. (And hopefully with a second human reviewer for the PR!)

I do think moving `fix = true` to the Ruff config makes sense, it'll be picked up when running Ruff as a CLI (outside pre-commit) too. I'll add this.

What do you think?

Reply:
Sure, I see the point about duplication. Looks pretty good to me now, thanks!